crates/sui/src/ptb/ptb_parser/parser.rs

```diff
@@ -1,93 +1,126 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
+
 use crate::ptb::ptb::PTBCommand;
+
 use move_command_line_common::{
     address::NumericalAddress,
     parser::{parse_u128, parse_u16, parse_u256, parse_u32, parse_u64, parse_u8, Parser, Token},
-    types::{ParsedType, TypeToken},
+    types::TypeToken,
 };
-use move_core_types::{identifier::Identifier, runtime_value::MoveValue};
+use move_core_types::identifier::Identifier;
 use std::{borrow::BorrowMut, marker::PhantomData, str::FromStr};
 
 use crate::ptb::ptb_parser::argument_token::ArgumentToken;
 use anyhow::{anyhow, bail, Context, Result};
 
-use super::command_token::CommandToken;
+use super::{
+    argument::Argument, command_token::CommandToken, context::PTBContext, errors::PTBError,
+};
 
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct ParsedPTBCommand<T> {
     pub name: CommandToken,
     pub args: Vec<T>,
 }
 
-impl ParsedPTBCommand<Argument> {
-    pub fn parse(cmd: &PTBCommand) -> Result<Self> {
-        let name = CommandToken::from_str(&cmd.name)?;
+pub struct ValueParser<
+    'a,
+    I: Iterator<Item = (ArgumentToken, &'a str)>,
+    P: BorrowMut<Parser<'a, ArgumentToken, I>>,
+> {
+    inner: P,
+    arg_index: usize,
+    _a: PhantomData<&'a ()>,
+    _i: PhantomData<I>,
+}
+
+pub struct PTBParser {
+    parsed: Vec<ParsedPTBCommand<Argument>>,
+    errors: Vec<PTBError>,
+    context: PTBContext,
+}
+
+impl PTBParser {
+    pub fn new() -> Self {
+        Self {
+            parsed: vec![],
+            errors: vec![],
+            context: PTBContext::new(),
+        }
+    }
+
+    pub fn finish(self) -> (Vec<ParsedPTBCommand<Argument>>, Vec<PTBError>) {
+        (self.parsed, self.errors)
+    }
+
+    pub fn parse(&mut self, mut cmd: PTBCommand) {
+        let name = CommandToken::from_str(&cmd.name);
+        if let Err(e) = name {
+            self.errors.push(PTBError::WithSource {
+                file_scope: self.context.current_file_scope().clone(),
+                message: format!("Failed to parse command name: {e}"),
+            });
+            return;
+        };
+        let name = name.unwrap();
+
         match name {
-            CommandToken::FileEnd
-            | CommandToken::FileStart
-            | CommandToken::Publish
-            | CommandToken::Upgrade => {
+            CommandToken::FileEnd => {
+                if let Err(e) = self.context.pop_file_scope(cmd.name.clone()) {
+                    self.errors.push(e);
+                }
+                return;
+            }
+            CommandToken::FileStart => {
+                let name = cmd.values.pop().unwrap();
+                self.context.push_file_scope(name);
+                return;
+            }
+            CommandToken::Publish | CommandToken::Upgrade => {
                 if cmd.values.len() != 1 {
```

Comment:
I think this will be caught by clap, but I prefer we're making sure we do something in this case.